### PR TITLE
Converted integration tests to unit tests

### PR DESCRIPTION
**Pull request by Zverik**
This pull request adds `unit_data.osm` and generates Gherkin files from it. The file is less than a megabyte and contains all data needed for testing. The whole cycle (running `unit.sh`) takes around 4 seconds on my machine — and that's with initializing the database! Fixed an error on the way.

### CODE REVIEW

File: tests/admin.sh [hunk 1]
```diff
@@ -0,0 +1,12 @@
+#!/bin/bash
+set -e -u
+HERE="$(dirname "$0")"
+VENV="${VENV:-$HERE/../.venv}"
```
Comment (gardster) [inline tests/admin.sh]: what script installs VENV?

File: tests/steps/requests.py [hunk 2]
```diff
@@ -13,3 +13,9 @@ def get_info(context, osm_type, osm_id):
     context.osm = (osm_type, osm_id)
     context.response = context.geocoder.get_info(osm_type[0], osm_id)
     context.response_code = context.geocoder.last_code
+
+
+@when('point is {point}')
+def find_point(context, point):
+    lonlat = context.points[point]
```
Comment (gardster) [inline tests/steps/requests.py]: optional:
` lon, lat = context.points[point]`

File: tests/unit.sh [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+fi
+
+# Run tests
+utils/osm_to_gherkin.py features_auto
+export PYTHONPATH="$HERE/tests"
+"$VENV/bin/behave" -f formatter:BareFormatter --tags unit ${TAGS-} $@ "$HERE/tests" || true
+rm -r features_auto
+
+# Stop web server
+kill "$FLASK_JOB"
```
Comment (gardster) [inline tests/unit.sh]: optional: You can bind kill on trap for cases if something went wrong inside the scripts.

File: tests/unit.sh [hunk 4]
```diff
@@ -0,0 +1,55 @@
+#!/bin/bash
+# This script runs unit tests on a pre-built OSM file.
```
Comment (gardster) [inline tests/unit.sh:2]: Script looks little bit inconsistent: it supposes that you have working PG database, have not uwsgi and have empty 80 port (you start uwsgi with empty config).
WDYT?

File: tests/utils/features_to_osm.py [hunk 5]
```diff
@@ -0,0 +1,86 @@
+#!/usr/bin/env python3
```
Comment (gardster) [inline tests/utils/features_to_osm.py]: Why do we need this file?

File: tests/utils/osm_to_gherkin.py [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+    ftags = scenarios[0].get('tags', set())
+    for scen in scenarios:
+        ftags = ftags.intersection(scen['tags'] or set())
+    ftags.add('@unit')
+    with open(filename, 'w') as f:
+        if ftags:
+            print(' '.join(ftags), file=f)
+        print(f'Feature: {feat.replace("_", " ").title()}', file=f)
+        print(file=f)
+        for scen in sorted(scenarios, key=lambda s: s.get('sid', 'zzz')):
```
Comment (gardster) [inline tests/utils/osm_to_gherkin.py]: why `zzz` is default?

File: tests/utils/osm_to_gherkin.py [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+            if 'tags' in scen and scen['tags'] - ftags:
+                print('    ' + ' '.join(scen['tags'] - ftags), file=f)
+            print(f'    Scenario: {scen["name"]}', file=f)
+            if 'note' in scen:
+                print('        ' + scen['note'], file=f)
+                print(file=f)
+            print(f'        When {scen["when"]}', file=f)
+            for k in (('then', 'Then'), ('and', 'And'), ('and1', 'And')):
```
Comment (gardster) [inline tests/utils/osm_to_gherkin.py]: What is `and1`? If we will have more checks will we add `and2`?

File: tests/utils/osm_to_gherkin.py [hunk 8]
```diff
@@ -0,0 +1,69 @@
+#!/usr/bin/env python3
```
Comment (gardster) [inline tests/utils/osm_to_gherkin.py:1]: Can you divide this file to several functions to make it more readable?

File: tests/utils/renumber_negative_ids.py [hunk 9]
```diff
@@ -0,0 +1,73 @@
+#!/usr/bin/env python3
+try:
+    from lxml import etree
+except ImportError:
+    import xml.etree.ElementTree as etree
+import sys
+
+if len(sys.argv) < 3:
```
Comment (gardster) [inline tests/utils/renumber_negative_ids.py:8]: Optional: maybe delegate all checks and help strings to argparse?

File: tests/utils/renumber_negative_ids.py [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+                print('Removing relation {}, since {} {} was not found.'.format(
+                    obj.get('id'), nd.get('type'), nd.get('ref')))
+                osm.remove(obj)
+                break
+            elif new_id > 0:
+                nd.set('ref', str(new_id))
+
+out_file = sys.argv[1] if sys.argv[2] == '-' else sys.argv[2]
+with open(out_file, 'wb') as f:
+    f.write(etree.tostring(osm, encoding='utf-8'))
```
Comment (gardster) [inline tests/utils/renumber_negative_ids.py]: optional `etree.write(f, encoding='utf'-8')`

**Review by gardster — CHANGES_REQUESTED** [review]

File: tests/admin.sh (hunk 1 above)
Comment (Zverik) [reply tests/admin.sh]: `run.sh` from the parent directory, and `unit.sh` from this one. Although you're right and it isn't obvious. I have added a section in README and rewritten this code to use `behave` from a user's virtualenv.

File: tests/steps/requests.py (hunk 2 above)
Comment (Zverik) [reply tests/steps/requests.py]: I wasn't sure this would work for a non-tuple. It does, so fixed. Thanks.

File: tests/unit.sh (hunk 3 above)
Comment (Zverik) [reply tests/unit.sh]: Done.

File: tests/unit.sh (hunk 4 above)
Comment (Zverik) [reply tests/unit.sh:2]: Not exactly. I start Flask development mode bypassing any wsgi. It uses port 5000, which is the default for the testing code (see `environment.py`). The need for PostgreSQL database is outlined in the README: cannot get around it.

File: tests/utils/features_to_osm.py (hunk 5 above)
Comment (Zverik) [reply tests/utils/features_to_osm.py]: I may need it for futher tests. But whatever: removed it to a local directory of mine.

File: tests/utils/osm_to_gherkin.py (hunk 6 above)
Comment (Zverik) [reply tests/utils/osm_to_gherkin.py]: To have `None` elements pushed to the bottom of the list.

File: tests/utils/osm_to_gherkin.py (hunk 7 above)
Comment (Zverik) [reply tests/utils/osm_to_gherkin.py]: Having more than three (`then` + `and` + `and1`) checks is extremely unlikely. But yes, if we would need a fourth, than that's an option.

File: tests/utils/osm_to_gherkin.py (hunk 8 above)
Comment (Zverik) [reply tests/utils/osm_to_gherkin.py:1]: Done.

File: tests/utils/renumber_negative_ids.py (hunk 9 above)
Comment (Zverik) [reply tests/utils/renumber_negative_ids.py:8]: Okay.

File: tests/utils/renumber_negative_ids.py (hunk 10 above)
Comment (Zverik) [reply tests/utils/renumber_negative_ids.py]: Okay.

**Comment by Zverik** [comment]
Pushed the changes, PTAL please.

**Review by gardster — APPROVED** [review]

**Pull request merged**